presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -124,6 +142,32 @@ public SqlStageExecution(StageStateMachine stateMachine, RemoteTaskFactory remot
             }
         }
         this.exchangeSources = fragmentToExchangeSource.build();
+
+        addStateChangeListener(newState -> {
+            if (!dynamicFilterSchedulingInfoPropagated &&
+                    (newState == SCHEDULING_SPLITS || newState == SCHEDULED || newState == RUNNING)) {
+                traverseNodesForDynamicFiltering(ImmutableList.of(stateMachine.getFragment().getRoot()));
+                dynamicFilterSchedulingInfoPropagated = true;
+            }
+        });
+    }
+
+    private void traverseNodesForDynamicFiltering(List<PlanNode> nodes)
```

Comment:
Do we know in planing phase what will be task IDs while execution? I think we don't.